Automattic/woocommerce-connect-client
# Add a "Phone Number" field to both the origin and destination address in the Labels UI

Fixes https://github.com/Automattic/woocommerce-connect-client/issues/574

You need to test it with the `add/phone_address` branch in the server.

**Background:**
EasyPost uses internally different APIs for ["regular" mail](https://www.usps.com/business/web-tools-apis/delivery-confirmation-domestic-shipping-label-api.htm) and for ["Priority Express" mail](https://www.usps.com/business/web-tools-apis/priority-mail-express-api.htm).

The normal API doesn't require the parameter `FromPhone` or `ToPhone`, but the Express API does (10-digit number, without punctuation). The solution is to add a `Phone number` field to the `Origin address` and `Destination address` form. This will make the phone field mandatory even for services that don't require it, but it's the simplest solution.

**Changes:**
* WooCommerce by default requires the _billing_ phone field to be filled, but there's no **shipping** phone. I've added that field.
* For orders done prior to this change, or for the ones that don't have an explicit *shipping address* (they only filled the Billing details) the phone number entered in the Billing Address is used instead.
* The cached merchant address for the Labels UI will no longer be valid (since it doesn't have a phone number), so the next time the merchant opens the Purchase Label UI he will have to input his phone (just once).
* The phone number validation is client-only. It uses [`libphonenumber`](https://github.com/seegno/google-libphonenumber) to pretty-print, normalize and validate the phone numbers.

**Possible things to improve**
* Right now the phone validator assumes that the phone number the merchant is entering will be of the same country as the address. This is a safe assumption for now, since the `USPS` API only accepts US-based phone numbers anyway. Do you think this could be a problem? Like, you're shipping from UK but your phone number is Canadian?
* The phone validation WooCommerce applies at checkout is really weak, it only checks that the entered string is a bunch of numbers. So it's possible that the merchant opens the Purchase Label UI and the entered phone is just garbage (and let's remember that's a required field). I guess that could also happen with an address...
* With this change, the phone number is required for every possible service. That means that if the merchant only wants to send items using `Priority Mail`, that field would not be needed at all. Maybe we should make it optional? Or maintain a list of services that require it? Or send the origin/destination phone along with the address in the `getRates` call and have some logic in the server to **not** return `Express` services if the phone is not present?

**To test**
* Buy something and go to the checkout.
* Check that you can fill the "Phone" field in the "Shipping Address".
* Go to the "Edit Order" page.
* Check that the phone number you entered appears on the Order Details.
* The "Origin Address" step should be marked as invalid, with the "Phone" field in red.
* Enter a phone number. Try with different phones, invalid phones, vanity phones, etc.
* Check that the phone you entered on the checkout appears in the "Destination Address" step.
* On the Rate selection step, choose a `Priority Express` service, and purchase the label(s).

@nabsul @allendav @robobot3000 @jeffstieler
